deltacat/compute/compactor/steps/repartition.py

```diff
@@ -0,0 +1,132 @@
+)
+from deltacat.utils.common import ReadKwargsProvider
+from deltacat.utils.performance import timed_invocation
+from deltacat.utils.metrics import emit_timer_metrics, MetricsConfig
+from deltacat.storage import Delta
+
+if importlib.util.find_spec("memray"):
+    import memray
+
+logger = logs.configure_deltacat_logger(logging.getLogger(__name__))
+
+
+def _timed_repartition(
+    annotated_delta: DeltaAnnotated,
+    repartition_type: str,
+    repartition_args: dict,
+    destination_partition: Partition,
+    enable_profiler: bool,
+    read_kwargs_provider: Optional[ReadKwargsProvider],
+    deltacat_storage=unimplemented_deltacat_storage,
+) -> RePartitionResult:
+
+    if repartition_type == "range":
+        # A delta that is partitioned by range is partitioned in such a way
+        # that each partition contains rows for which the partitioning expression value lies within a given range for a specified column.
+        # For example, if the partitioning column is "date", the partition ranges are "2020-01-01" to "2020-01-10" and "2020-01-11" to "2020-01-20".
+        # The partitioned delta will have two partitions, one for each range.
+        column: str = repartition_args["column"]
+        partition_ranges: List = repartition_args["ranges"]
+        task_id = get_current_ray_task_id()
+        worker_id = get_current_ray_worker_id()
+        with memray.Tracker(
+            f"repartition_{worker_id}_{task_id}.bin"
+        ) if enable_profiler else nullcontext():
+            tables: List[pa.Table] = deltacat_storage.download_delta(
+                annotated_delta,
+                storage_type=StorageType.LOCAL,
+                file_reader_kwargs_provider=read_kwargs_provider,
+            )
+            partitioned_tables = {i: [] for i in range(len(partition_ranges) + 1)}
+            total_record_count = 0
+            col_name_int64 = f"{column}_int64"
+            for table in tables:
+                total_record_count += len(table)
+                table_new = table.add_column(
+                    0,
+                    pa.field(col_name_int64, pa.int64()),
+                    pc.cast(table[column], pa.int64()),
+                )
+                for filter_value in partition_ranges:
+                    partitioned_tables[partition_ranges.index(filter_value)].append(
+                        table_new.filter(
+                            (pc.field(col_name_int64) <= pc.scalar(filter_value))
+                        )
+                    )
+                    if filter_value == partition_ranges[-1]:
+                        partitioned_tables[partition_ranges.index(filter_value)].append(
+                            table_new.filter(
+                                (pc.field(col_name_int64) > pc.scalar(filter_value))
+                            )
+                        )
+            # TODO(rootliu) set optimal or max number of records per file to defer the performance degradation due to too many small files
```

Comment:
This can be configurable and optimality decision can be delegated to caller.

Reply:
I think this needs to be part of the first release itself. We won't get enough time to work on it after 6/30 as this approach would end up creating lot of small files defeating the purpose of compaction.